gnocchi/rest/api.py

```diff
@@ -1890,6 +1899,136 @@ class BatchController(object):
     resources = ResourcesBatchController()
 
 
+class PrometheusWriteController(rest.RestController):
+
+    PROMETHEUS_RESOURCE_TYPE = {
+        "instance": {"type": "string",
+                     "min_length": 1,
+                     "max_length": 512,
+                     "required": True},
+        "job": {"type": "string",
+                "min_length": 1,
+                "max_length": 512,
+                "required": True}
+    }
+
+    # Retry with exponential backoff for up to 1 minute
+    @classmethod
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=0.5, max=60),
+        retry=tenacity.retry_if_exception_type(
+            (indexer.NoSuchResource, indexer.ResourceAlreadyExists,
+             indexer.ResourceTypeAlreadyExists,
+             indexer.NamedMetricAlreadyExists)))
+    def get_or_create_resource_and_metrics(cls, creator, rid,
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
+        except indexer.NoSuchResourceType:
+            enforce("create resource type", {
+                'name': 'prometheus',
+                'state': 'creating',
+                'attributes': cls.PROMETHEUS_RESOURCE_TYPE
+            })
+
+            schema = pecan.request.indexer.get_resource_type_schema()
+            rt = schema.resource_type_from_dict(
+                'prometheus', cls.PROMETHEUS_RESOURCE_TYPE, 'creating')
+            pecan.request.indexer.create_resource_type(rt)
```

Comment:
can this throw an exception too? if  say two things try to create same resource_type at same time?

Reply:
IIUC this can raise `indexer.ResourceTypeAlreadyExists` which is gonna make the code retry (see the tenacity decorator).